### ISSUE TITLE
v1: How to handle the Cedar instance's relationship to the DOM node and chart instance?

### ISSUE THREAD

**Issue by tomwayson** [Esri/cedar#297]
Right now, every time someone wants to update a chart, they have to call `show()` w/ the element id. Consumers should not need hold on to an id and pass it in every time - that should be a one time thing.

So, do we mirror the v0 API and make `show()` a thin wrapper that simply sets the id property on the instance and then calls `update()`? Certain aspects of that API have always bothered me. The fact that the element id is passed to `show()` instead of the constructor _implies_ that the instance is not associated w/ a particular DOM node, but it definitely is. Not to mention the confusing way `show()` takes other options that [simply override instance properties](https://github.com/Esri/cedar/blob/c793cd4ae8108798365b9f8ff0decb0791c36986/src/cedar.js#L382-L398). If you call `.show()` again w/ a different id and options I think you will see two charts that differ only in the options you passed (i.e same type & data). However I think the first has become a zombie chart because if you then went and modified the `datasets` and called `update()` I think only the second one will update. I don't even want to think about what happens to the events in that scenario.

All in all I think that API is broken and that we should come up w/ something different. I like the idea of passing the element id as a required parameter to the constructor - that firmly establishes the relationship of the instance to a particular element. However, beyond that, I'm not sure what to do. If the constructor takes the element id, there's pretty much no difference between `show()` and `update()`. If you only have on function to fetch the data and render the chart, what do you call it? `let chart = new Cedar(config); chart.update();` feels a bit odd. Technically you could have `new Cedar(config)` kick off the update process, but I'm not a fan of async side effects (>cough< `new Map()`). `chart.render()`? That doesn't really capture the notion that we're fetching data. Maybe we need separate methods for fetching the data and rendering the chart and a wrapper that calls them both. I like that idea, but what do we name each of those functions?

Also In the long run we might not want to create a new chart every time the "render" method is called. See [this AmCharts example](https://www.amcharts.com/kbase/dynamically-loading-chart-datasets/). So we may want to start holding on to the chart instance itself too.

**Comment by benstoltz** [comment]
@tomwayson @ajturner I like the notion of splitting `.show` into data fetch and rendering methods and then have `.show` call them both like a wrapper. We could call the fetch `.getData` and the render `.render`. Maybe?

**Comment by tomwayson** [comment]
Thanks @benstoltz 

In slack, @benstoltz and I also decided to not have the constructor actually try show the chart even though if it has all the things it needs.

Based on that (and the consensus on the names for global and constructor in #279) I propose this as the API (silly me for even trying to have the above API discussion w/o pseudo code):

```
// to create the chart and show it for the first time
var chart = new cedar.Chart(elementId, options);
chart.show();
// to update the chart after changing chart properties or refreshing data
chart.show(); // the name show() feels a little awkward in the update context
```

Works for me. I'm not super keen on `show()` in the context of an update, but the name `update()` would not work at all in the context of having just created the chart. I'm not sure we're going to find a name that works better than `show()` for both contexts.

**Comment by ajturner** [comment]
`chart.display();`

Or

`chart.render();`

Or

`chart.draw();`

On Oct 9, 2017, at 10:59 AM, Tom Wayson <notifications@github.com<mailto:notifications@github.com>> wrote:


Thanks @benstoltz<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_benstoltz&d=DwMFaQ&c=n6-cguzQvX_tUIrZOS_4Og&r=k9u7PfHvjHW9lzI6dYwFzg&m=Lie5h0cMCUXFlh1tl3TLsto1bFN1uyyLwPU-sUktLrA&s=af6bWaNvXhfUZ45LA4w55Ny41wzMNd_hAm8i_HZaRfU&e=>

In slack, @benstoltz<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_benstoltz&d=DwMFaQ&c=n6-cguzQvX_tUIrZOS_4Og&r=k9u7PfHvjHW9lzI6dYwFzg&m=Lie5h0cMCUXFlh1tl3TLsto1bFN1uyyLwPU-sUktLrA&s=af6bWaNvXhfUZ45LA4w55Ny41wzMNd_hAm8i_HZaRfU&e=> and I also decided to not have the constructor actually try show the chart even though if it has all the things it needs.

Based on that (and the consensus on the names for global and constructor in #279<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_Esri_cedar_issues_279&d=DwMFaQ&c=n6-cguzQvX_tUIrZOS_4Og&r=k9u7PfHvjHW9lzI6dYwFzg&m=Lie5h0cMCUXFlh1tl3TLsto1bFN1uyyLwPU-sUktLrA&s=nhvEbjTy4MyoJ9sxDJxPkQ-P1gS6mYJZALF-E2Yktqw&e=>) I propose this as the API (silly me for even trying to have the above API discussion w/o pseudo code):

// to create the chart and show it for the first time
var chart = new cedar.Chart(elementId, options);
chart.show();
// to update the chart after changing chart properties or refreshing data
chart.show(); // the name show() feels a little awkward in the update context


Works for me. I'm not super keen on show() in the context of an update, but the name update() would not work at all in the context of having just created the chart. I'm not sure we're going to find a name that works better than show() for both contexts.

**Comment by tomwayson** [comment]
Resolved in #308
